File: src/Flagrum.Blender/core/graphics/materials/gamematerial.py
from dataclasses import dataclass

from .buffer import Buffer
from .sampler import Sampler
from .shaderbinary import ShaderBinary
from .shaderprogram import ShaderProgram
from .texture import Texture
from .uniform import Uniform
from ..graphicsbinary import GraphicsBinary
from ...serialization.messagepack.writer import MessagePackWriter
from ...serialization.stringbuffer import StringBuffer
from ...serialization.valuebuffer import ValueBuffer
from ...utilities.cryptography import Cryptography


@dataclass
class GameMaterial(GraphicsBinary):
    name_offset: int
    unknown: int
    uniform_count: int
    buffer_count: int
    texture_count: int
    sampler_count: int
    total_uniform_count: int
    total_buffer_count: int
    total_texture_count: int
    total_sampler_count: int
    shader_binary_count: int
    shader_program_count: int
    value_buffer_size: int
    string_buffer_size: int
    name_hash: int
    unknown_hash: int
    blend_type: int
    blend_factor: float
    render_state_bits: int
    skin_vs_max_bone_count: int
    brdf_type: int
    htpk_uri_offset: int
    uniforms: list[Uniform]
    buffers: list[Buffer]
    textures: list[Texture]
    samplers: list[Sampler]
    shader_binaries: list[ShaderBinary]
    shader_programs: list[ShaderProgram]
    value_buffer: ValueBuffer
    string_buffer: StringBuffer

    name: str
    htpk_uri: str

    def __init__(self, reader):
        GraphicsBinary.__init__(self)
        GraphicsBinary.read(self, reader)

        self.name_offset = reader.read()

        if self.version >= 20220707:
            self.unknown = reader.read()
        else:
            self.unknown = 0

        self.buffer_count = reader.read()
        self.uniform_count = reader.read()
        self.texture_count = reader.read()
        self.sampler_count = reader.read()
        self.total_buffer_count = reader.read()
        self.total_uniform_count = reader.read()
        self.total_texture_count = reader.read()
        self.total_sampler_count = reader.read()
        self.shader_binary_count = reader.read()
        self.shader_program_count = reader.read()
        self.value_buffer_size = reader.read()
        self.string_buffer_size = reader.read()
        self.name_hash = reader.read()

        if self.version >= 20220707:
            self.unknown_hash = reader.read()
        else:
            self.unknown_hash = 0

        self.blend_type = reader.read()
        self.blend_factor = reader.read()
        self.render_state_bits = reader.read()
        self.skin_vs_max_bone_count = reader.read()
        self.brdf_type = reader.read()
        self.htpk_uri_offset = reader.read()

        self.buffers = []
        for i in range(self.total_buffer_count):
            self.buffers.append(Buffer(reader))

        self.uniforms = []
        for i in range(self.total_uniform_count):
            self.uniforms.append(Uniform(reader))

        self.textures = []
        for i in range(self.total_texture_count):
            self.textures.append(Texture(reader))

        self.samplers = []
        for i in range(self.total_sampler_count):
            self.samplers.append(Sampler(reader))

        self.shader_binaries = []
        for i in range(self.shader_binary_count):
            self.shader_binaries.append(ShaderBinary(reader))

        self.shader_programs = []
        for i in range(self.shader_program_count):
            self.shader_programs.append(ShaderProgram(reader))

        self.value_buffer = ValueBuffer()
        self.value_buffer.initialize(reader.read())
        self.string_buffer = StringBuffer()
        self.string_buffer.initialize(reader.read())

        self.name = self.string_buffer.get(self.name_offset)
        self.htpk_uri = self.string_buffer.get(self.htpk_uri_offset)

        for i in range(self.total_buffer_count):
            self.buffers[i].name = self.string_buffer.get(self.buffers[i].name_offset)
            self.buffers[i].shader_gen_name = self.string_buffer.get(self.buffers[i].shader_gen_name_offset)
            if self.buffers[i].uniform_index < self.uniform_count:
                self.buffers[i].values = self.value_buffer.get(self.buffers[i].offset, self.buffers[i].size)

        for i in range(self.total_uniform_count):
            self.uniforms[i].name = self.string_buffer.get(self.uniforms[i].name_offset)
            self.uniforms[i].shader_gen_name = self.string_buffer.get(self.uniforms[i].shader_gen_name_offset)
            if i < self.uniform_count:
                self.uniforms[i].values = self.value_buffer.get(self.uniforms[i].offset, self.uniforms[i].size)

        for i in range(self.total_texture_count):
            self.textures[i].name = self.string_buffer.get(self.textures[i].name_offset)
            self.textures[i].shader_gen_name = self.string_buffer.get(self.textures[i].shader_gen_name_offset)
            self.textures[i].uri = self.string_buffer.get(self.textures[i].uri_offset)

        for i in range(self.total_sampler_count):
            self.samplers[i].name = self.string_buffer.get(self.samplers[i].name_offset)
            self.samplers[i].shader_gen_name = self.string_buffer.get(self.samplers[i].shader_gen_name_offset)

        for i in range(self.shader_binary_count):
            self.shader_binaries[i].uri = self.string_buffer.get(self.shader_binaries[i].uri_offset)

    def write(self, writer: MessagePackWriter):
        self.update_value_buffer()
        self.update_string_buffer()
        self.regenerate_dependency_table()

        GraphicsBinary.write(self, writer)

        writer.write(self.name_offset)
        writer.write(self.buffer_count)
        writer.write(self.uniform_count)
        writer.write(self.texture_count)
        writer.write(self.sampler_count)
        writer.write(self.total_buffer_count)
        writer.write(self.total_uniform_count)
        writer.write(self.total_texture_count)
        writer.write(self.total_sampler_count)
        writer.write(self.shader_binary_count)
        writer.write(self.shader_program_count)
        writer.write(self.value_buffer_size)
        writer.write(self.string_buffer_size)
        writer.write(self.name_hash)
        writer.write(self.blend_type)
        writer.write(self.blend_factor)
        writer.write(self.render_state_bits)
        writer.write(self.skin_vs_max_bone_count)
        writer.write(self.brdf_type)
        writer.write(self.htpk_uri_offset)

        for buffer in self.buffers:
            buffer.write(writer)

        for uniform in self.uniforms:
            uniform.write(writer)

        for texture in self.textures:
            texture.write(writer)

        for sampler in self.samplers:
            sampler.write(writer)

        for shader_binary in self.shader_binaries:
            shader_binary.write(writer)

        for shader_program in self.shader_programs:
            shader_program.write(writer)

        writer.write(self.value_buffer.to_bytes())
        writer.write(self.string_buffer.to_bytes())

    def update_value_buffer(self):
        value_buffer = ValueBuffer()

        for buffer in self.buffers:
            if buffer.uniform_index < self.uniform_count:
                buffer.offset = value_buffer.put(buffer.values)

        for i in range(self.uniform_count):
            uniform = self.uniforms[i]
            uniform.offset = value_buffer.put(uniform.values)

        self.value_buffer = value_buffer
        self.value_buffer_size = value_buffer.size()

    def update_string_buffer(self):
        string_buffer = StringBuffer()

        self.name_offset = string_buffer.put(self.name)

        for buffer in self.buffers:
            buffer.name_offset = string_buffer.put(buffer.name)
            buffer.shader_gen_name_offset = string_buffer.put(buffer.shader_gen_name)

        for uniform in self.uniforms:
            uniform.name_offset = string_buffer.put(uniform.name)
            uniform.shader_gen_name_offset = string_buffer.put(uniform.shader_gen_name)

        for texture in self.textures:
            texture.name_offset = string_buffer.put(texture.name)
            texture.shader_gen_name_offset = string_buffer.put(texture.shader_gen_name)
            texture.uri_offset = string_buffer.put(texture.uri)

        for sampler in self.samplers:
            sampler.name_offset = string_buffer.put(sampler.name)
            sampler.shader_gen_name_offset = string_buffer.put(sampler.shader_gen_name)

        for binary in self.shader_binaries:
            binary.uri_offset = string_buffer.put(binary.uri)

        self.htpk_uri_offset = string_buffer.put(self.htpk_uri)

        self.string_buffer = string_buffer
        self.string_buffer_size = string_buffer.size()

    def regenerate_dependency_table(self):
        dependencies = {}
        self.hashes = []
        uri = self.dependencies["ref"]
        asset_uri = uri[:(uri.rindex("/") + 1)]

        for binary in self.shader_binaries:
            if binary.uri_hash > 0:
                dependencies[str(binary.uri_hash)] = binary.uri
                self.hashes.append(binary.uri_hash)

        for texture in self.textures:
            if texture.uri_hash > 0:
                dependencies[str(texture.uri_hash)] = texture.uri
                self.hashes.append(texture.uri_hash)

        dependencies["asset_uri"] = asset_uri
        dependencies["ref"] = uri

        self.dependencies = dependencies

    def get_buffer(self, shader_gen_name: str):
        for i in range(len(self.buffers)):
            if self.buffers[i].shader_gen_name is not None \
                    and self.buffers[i].shader_gen_name.upper() == shader_gen_name.upper():
                return self.buffers[i]
        return None

    def set_texture_uri(self, texture_slot: str, uri: str):
        for texture in self.textures:
            if texture.shader_gen_name.upper() == texture_slot.upper():
                texture.uri = uri
                texture.uri_hash32 = Cryptography.hash32(uri)
                texture.uri_hash = Cryptography.uri_hash64(uri)
                break

    def set_buffer_value(self, buffer_name: str, value: list[float]):
        for buffer in self.buffers:
            if buffer.shader_gen_name.upper() == buffer_name.upper():
                buffer.values = value

File: src/Flagrum.Core/Graphics/Materials/GameMaterial.cs
using System.Collections.Generic;
using System.IO;
using System.Linq;
using Flagrum.Core.Graphics.Containers;
using Flagrum.Core.Serialization;
using Flagrum.Core.Serialization.MessagePack;

namespace Flagrum.Core.Graphics.Materials;

public class GameMaterial : GraphicsBinary
{
    public ulong NameOffset { get; set; }
    public ushort Unknown { get; set; }

    // These are the counts of the objects in the lists that map to this material
    public ushort UniformCount { get; set; }
    public ushort BufferCount { get; set; }
    public ushort TextureCount { get; set; }
    public ushort SamplerCount { get; set; }

    // These are the counts of all objects in the lists
    public ushort TotalUniformCount { get; set; }
    public ushort TotalBufferCount { get; set; }
    public ushort TotalTextureCount { get; set; }
    public ushort TotalSamplerCount { get; set; }

    public ushort ShaderBinaryCount { get; set; }
    public ushort ShaderProgramCount { get; set; }

    public uint ValueBufferSize { get; set; }
    public uint StringBufferSize { get; set; }

    public uint NameHash { get; set; }
    public uint Unknown2 { get; set; }
    public uint BlendType { get; set; }
    public float BlendFactor { get; set; }
    public uint RenderStateBits { get; set; }
    public ushort SkinVsMaxBoneCount { get; set; }
    public ushort BrdfType { get; set; }
    public ulong HighTexturePackUriOffset { get; set; }

    public ValueBuffer ValueBuffer { get; set; }
    public StringBuffer StringBuffer { get; set; }

    public List<GameMaterialUniform> Uniforms { get; set; } = new();
    public List<GameMaterialBuffer> Buffers { get; set; } = new();
    public List<GameMaterialTexture> Textures { get; set; } = new();
    public List<GameMaterialSampler> Samplers { get; set; } = new();
    public List<GameMaterialShaderBinary> ShaderBinaries { get; set; } = new();
    public List<GameMaterialShaderProgram> ShaderPrograms { get; set; } = new();

    public string Name { get; set; }
    public string HighTexturePackUri { get; set; }

    public static GameMaterial Deserialize(byte[] buffer)
    {
        var material = new GameMaterial();
        material.Read(buffer);
        return material;
    }

    public override void Read(Stream stream)
    {
        base.Read(stream);

        using var reader = new MessagePackReader(stream);
        reader.DataVersion = Version;

        NameOffset = reader.Read<ulong>();

        if (reader.DataVersion >= 20220707)
        {
            Unknown = reader.Read<ushort>();
        }

        BufferCount = reader.Read<ushort>();
        UniformCount = reader.Read<ushort>();
        TextureCount = reader.Read<ushort>();
        SamplerCount = reader.Read<ushort>();
        TotalBufferCount = reader.Read<ushort>();
        TotalUniformCount = reader.Read<ushort>();
        TotalTextureCount = reader.Read<ushort>();
        TotalSamplerCount = reader.Read<ushort>();
        ShaderBinaryCount = reader.Read<ushort>();
        ShaderProgramCount = reader.Read<ushort>();
        ValueBufferSize = reader.Read<uint>();
        StringBufferSize = reader.Read<uint>();
        NameHash = reader.Read<uint>();

        if (reader.DataVersion >= 20220707)
        {
            Unknown2 = reader.Read<uint>();
        }

        BlendType = reader.Read<uint>();
        BlendFactor = reader.Read<float>();
        RenderStateBits = reader.Read<uint>();
        SkinVsMaxBoneCount = reader.Read<ushort>();
        BrdfType = reader.Read<ushort>();

        if (reader.DataVersion >= 20150403)
        {
            HighTexturePackUriOffset = reader.Read<ushort>();
        }

        for (var i = 0; i < TotalBufferCount; i++)
        {
            var buffer = new GameMaterialBuffer();
            buffer.Read(reader);
            Buffers.Add(buffer);
        }

        for (var i = 0; i < TotalUniformCount; i++)
        {
            var uniform = new GameMaterialUniform();
            uniform.Read(reader);
            Uniforms.Add(uniform);
        }

        for (var i = 0; i < TotalTextureCount; i++)
        {
            var texture = new GameMaterialTexture();
            texture.Read(reader);
            Textures.Add(texture);
        }

        for (var i = 0; i < TotalSamplerCount; i++)
        {
            var sampler = new GameMaterialSampler();
            sampler.Read(reader);
            Samplers.Add(sampler);
        }

        for (var i = 0; i < ShaderBinaryCount; i++)
        {
            var shaderBinary = new GameMaterialShaderBinary();
            shaderBinary.Read(reader);
            ShaderBinaries.Add(shaderBinary);
        }

        for (var i = 0; i < ShaderProgramCount; i++)
        {
            var shaderProgram = new GameMaterialShaderProgram();
            shaderProgram.Read(reader);
            ShaderPrograms.Add(shaderProgram);
        }

        if (ValueBufferSize > 0)
        {
            ValueBuffer = new ValueBuffer(reader.Read<byte[]>());
        }

        StringBuffer = new StringBuffer(reader.Read<byte[]>());

        // Read values from the value buffer
        foreach (var buffer in Buffers)
        {
            if (buffer.UniformIndex < UniformCount)
            {
                var floatCount = buffer.Size / 4;
                if (floatCount > 0)
                {
                    buffer.Values = new float[floatCount];
                    for (var i = 0u; i < floatCount; i++)
                    {
                        buffer.Values[i] = ValueBuffer.Get(buffer.Offset + i * 4);
                    }
                }
            }
        }

        // Read strings from the string buffer
        Name = StringBuffer.Get(NameOffset);
        HighTexturePackUri = StringBuffer.Get(HighTexturePackUriOffset);

        for (var i = 0; i < TotalBufferCount; i++)
        {
            Buffers[i].Name = StringBuffer.Get(Buffers[i].NameOffset);
            Buffers[i].ShaderGenName = StringBuffer.Get(Buffers[i].ShaderGenNameOffset);
        }

        for (var i = 0; i < TotalUniformCount; i++)
        {
            Uniforms[i].Name = StringBuffer.Get(Uniforms[i].NameOffset);
            Uniforms[i].ShaderGenName = StringBuffer.Get(Uniforms[i].ShaderGenNameOffset);
        }

        for (var i = 0; i < TotalTextureCount; i++)
        {
            Textures[i].Name = StringBuffer.Get(Textures[i].NameOffset);
            Textures[i].ShaderGenName = StringBuffer.Get(Textures[i].ShaderGenNameOffset);
            Textures[i].Uri = StringBuffer.Get(Textures[i].UriOffset);
        }

        for (var i = 0; i < TotalSamplerCount; i++)
        {
            Samplers[i].Name = StringBuffer.Get(Samplers[i].NameOffset);
            Samplers[i].ShaderGenName = StringBuffer.Get(Samplers[i].ShaderGenNameOffset);
        }

        for (var i = 0; i < ShaderBinaryCount; i++)
        {
            ShaderBinaries[i].Uri = StringBuffer.Get(ShaderBinaries[i].UriOffset);
        }
    }

    public override void Write(Stream stream)
    {
        UpdateStringBuffer();
        RegenerateDependencyTable();

        base.Write(stream);

        using var writer = new MessagePackWriter(stream);
        writer.DataVersion = Version;

        writer.Write(NameOffset);

        if (writer.DataVersion >= 20220707)
        {
            writer.Write(Unknown);
        }

        writer.Write(BufferCount);
        writer.Write(UniformCount);
        writer.Write(TextureCount);
        writer.Write(SamplerCount);
        writer.Write(TotalBufferCount);
        writer.Write(TotalUniformCount);
        writer.Write(TotalTextureCount);
        writer.Write(TotalSamplerCount);
        writer.Write(ShaderBinaryCount);
        writer.Write(ShaderProgramCount);
        writer.Write(ValueBufferSize);
        writer.Write(StringBufferSize);
        writer.Write(NameHash);

        if (writer.DataVersion >= 20220707)
        {
            writer.Write(Unknown2);
        }

        writer.Write(BlendType);
        writer.Write(BlendFactor);
        writer.Write(RenderStateBits);
        writer.Write(SkinVsMaxBoneCount);
        writer.Write(BrdfType);

        if (writer.DataVersion >= 20150403)
        {
            writer.Write(HighTexturePackUriOffset);
        }

        foreach (var buffer in Buffers)
        {
            buffer.Write(writer);
        }

        foreach (var uniform in Uniforms)
        {
            uniform.Write(writer);
        }

        foreach (var texture in Textures)
        {
            texture.Write(writer);
        }

        foreach (var sampler in Samplers)
        {
            sampler.Write(writer);
        }

        foreach (var shaderBinary in ShaderBinaries)
        {
            shaderBinary.Write(writer);
        }

        foreach (var shaderProgram in ShaderPrograms)
        {
            shaderProgram.Write(writer);
        }

        writer.Write(ValueBuffer.ToArray());
        writer.Write(StringBuffer.ToArray());
    }

    public void RegenerateDependencyTable()
    {
        var dependencies = new Dictionary<string, string>();
        var uri = Dependencies["ref"];
        var assetUri = uri.Remove(uri.LastIndexOf('/') + 1);

        foreach (var shaderBinary in ShaderBinaries.Where(s => s.UriHash > 0))
        {
            dependencies[shaderBinary.UriHash.ToString()] = shaderBinary.Uri;
        }

        foreach (var texture in Textures.Where(t => t.UriHash > 0))
        {
            dependencies[texture.UriHash.ToString()] = texture.Uri;
        }

        dependencies["asset_uri"] = assetUri;
        dependencies["ref"] = uri;

        Dependencies = dependencies;
        Hashes = dependencies
            .Where(d => ulong.TryParse(d.Key, out _))
            .Select(d => ulong.Parse(d.Key))
            .OrderBy(h => h)
            .ToList();
    }

    private void UpdateStringBuffer()
    {
        StringBuffer = new StringBuffer();
        NameOffset = StringBuffer.Put(Name);

        foreach (var buffer in Buffers)
        {
            buffer.NameOffset = StringBuffer.Put(buffer.Name);
            buffer.ShaderGenNameOffset = StringBuffer.Put(buffer.ShaderGenName);
        }

        foreach (var uniform in Uniforms)
        {
            uniform.NameOffset = StringBuffer.Put(uniform.Name);
            uniform.ShaderGenNameOffset = StringBuffer.Put(uniform.ShaderGenName);
        }

        foreach (var texture in Textures)
        {
            texture.NameOffset = StringBuffer.Put(texture.Name);
            texture.ShaderGenNameOffset = StringBuffer.Put(texture.ShaderGenName);
            texture.UriOffset = StringBuffer.Put(texture.Uri);
        }

        foreach (var sampler in Samplers)
        {
            sampler.NameOffset = StringBuffer.Put(sampler.Name);
            sampler.ShaderGenNameOffset = StringBuffer.Put(sampler.ShaderGenName);
        }

        foreach (var binary in ShaderBinaries)
        {
            binary.UriOffset = StringBuffer.Put(binary.Uri);
        }

        HighTexturePackUriOffset = StringBuffer.Put(HighTexturePackUri);
        StringBufferSize = (uint)StringBuffer.Length;
    }

    public void UpdateValueBuffer()
    {
        ValueBuffer = new ValueBuffer(ValueBufferSize);

        for (var i = 0; i < TotalUniformCount; i++)
        {
            var uniform = Uniforms[i];

            if ((uniform.Flags & 1) == 0)
            {
                continue;
            }

            foreach (var buffer in Buffers.Where(b => b.UniformIndex == i))
            {
                ValueBuffer.Put(Uniforms[buffer.UniformIndex].Offset + buffer.Offset, buffer.Values);
            }
        }
    }
}